plugins/amd-kria/fu-amd-kria-plugin.c

```diff
@@ -0,0 +1,127 @@
+ * AMD Chooses the MIT license part of Dual MIT/LGPLv2 license agreement.
+ *
+ * SPDX-License-Identifier: LGPL-2.1-or-later OR MIT
+ */
+
+#include "config.h"
+
+#include "fu-amd-kria-plugin.h"
+#include "fu-amd-kria-device.h"
+#include "fu-amd-kria-image-reg.h"
+#include "fu-amd-kria-persistent-reg.h"
+
+struct _FuAmdKriaPlugin {
+	FuPlugin parent_instance;
+};
+
+G_DEFINE_TYPE(FuAmdKriaPlugin, fu_amd_kria_plugin, FU_TYPE_PLUGIN)
+
+static void
+fu_amd_kria_plugin_init(FuAmdKriaPlugin *self)
+{
+}
+
+static gboolean
+fu_amd_kria_plugin_process_image(FuPlugin *plugin, FuDevice *dev, GError **error)
+{
+	FuDevice *system;
+	g_autoptr(GBytes) bytes = NULL;
+	g_autoptr(FuFirmware) firmware = NULL;
+	g_autoptr(FuProgress) progress = fu_progress_new(NULL);
+
+	bytes = fu_device_dump_firmware(dev, progress, error);
+	if (bytes == NULL)
+		return FALSE;
+	firmware = fu_amd_kria_image_firmware_new();
+
+	if (!fu_firmware_parse(firmware, bytes, 0x0, error))
+		return FALSE;
+
+	system = fu_plugin_cache_lookup(plugin, "system-firmware");
+	if (system != NULL)
+		fu_device_set_version(system, fu_firmware_get_version(firmware));
+
+	fu_device_set_version(dev, fu_firmware_get_version(firmware));
+
+	return TRUE;
+}
+
+static void
+fu_amd_kria_plugin_device_registered(FuPlugin *plugin, FuDevice *dev)
+{
+	g_autoptr(GError) error_local = NULL;
+
+	if (g_strcmp0(fu_device_get_plugin(dev), "mtd") != 0)
+		return;
+
+	//TODO; need to resolve if pulling from Image A/Image B and figuring out active makes sense
+	if (g_strcmp0(fu_device_get_name(dev), "Image A") == 0 ||
+	    g_strcmp0(fu_device_get_name(dev), "Image B") == 0) {
+		if (!fu_amd_kria_plugin_process_image(plugin, dev, &error_local))
+			g_warning("%s", error_local->message);
+	}
+
+	fu_device_remove_flag(dev, FWUPD_DEVICE_FLAG_UPDATABLE);
+	fu_plugin_cache_add(plugin, fu_device_get_name(dev), dev);
+}
+
+static gboolean
+fu_amd_kria_plugin_coldplug(FuPlugin *plugin, FuProgress *progress, GError **error)
+{
+	FuContext *ctx = fu_plugin_get_context(plugin);
+	GPtrArray *hwids = fu_context_get_hwid_guids(ctx);
+	g_autofree gchar *sysfsfwdir = fu_path_from_kind(FU_PATH_KIND_SYSFSDIR_FW);
+	g_autofree gchar *esrt_path = g_build_filename(sysfsfwdir, "efi", "esrt", NULL);
+	g_autofree gchar *vendor_id = NULL;
+	g_autoptr(FuDevice) dev = NULL;
+
+	/* If there is an ESRT use that instead and disable the plugin */
+	if (g_file_test(esrt_path, G_FILE_TEST_IS_DIR)) {
+		fu_plugin_add_flag(plugin, FWUPD_PLUGIN_FLAG_DISABLED);
+		return TRUE;
+	}
+
+	dev = g_object_new(FU_TYPE_AMD_KRIA_DEVICE,
+			   "context",
+			   fu_plugin_get_context(plugin),
+			   NULL);
+
+	/* these come from FDT */
+	vendor_id = g_strdup_printf("DMI:%s", fu_context_get_hwid_value(ctx, FU_HWIDS_KEY_MANUFACTURER));
+	fu_device_add_vendor_id(dev, vendor_id);
+	for (guint i = 0; i < hwids->len; i++) {
+		const gchar *hwid = g_ptr_array_index(hwids, i);
+		fu_device_add_guid(dev, hwid);
```

Comment:
using the "raw" GUID here seems risky. Might it bet better to *build* an instance ID with the GUID, e.g. `MTD/HWID_%s&NAME_%s` or something like that?

Reply:
Agree.

I was talking to the team about it, and there is an I2C EEPROM that in this generation has a string we'll use to build GUID and in next generation will have a GUID included.  I'm going to switch over to that instead.

Reply:
Fixed to now be an a derived I2C plugin that gets the EEPROM data to build all the interesting stuff.